### ISSUE TITLE
Autosubmit Job List Structure persistence

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#511]
In GitLab by @wuruchi on Mar 26, 2020, 14:45

Hello @mcastril @dbeltrankyl 

I am creating this issue to help track the progress of this new feature and post any doubts that appear about which way its development should go. In addition, it will serve as a source of information for future documentation.

## Problem:

As we know, the `transitive reduction` of the graph generated from the experiment has the highest time and computation cost of all the processes involved in JobList creation and management. Furthermore, JobList related processes are constantly executed in the lifecycle of an experiment.

## Objective:

The main objective of this feature is to offer the user the possibility to save the result of the transitive reduction so Autosubmit can access it and avoid performing `transitive reduction` every time it needs to read the JobList. The bigger, in the number of jobs, the experiment is, the greater the time and computation cost saved.

### Secondary Objective:

Develop and present a historic complexity analysis of experiments managed by Autosubmit. 

## Analysis:

It seems interesting and useful to take a look at the experiments Autosubmit has been managing so far. Here is some analysis based on the data Autosubmit API has been collecting in the past weeks:

### Overview

There are a total of 2778 unique registered experiments. After we filter out those experiments with 0 jobs or those in which less than 10% of jobs have been completed, we end up with **2038** *valid* experiments.

In the following plot, we see the distribution of experiments by their number of jobs.

![experiments](https://earth.bsc.es/gitlab/es/autosubmit/uploads/b0514ad762c29f847d5ad7ffb4542017/experiments.jpg)

Where $`\mu = 759`$ and $`\overline{x} = 56`$.

The time cost of the transitive reduction process, which is executed regularly while using the features of Autosubmit, has polynomial growth as a function of the number of jobs $`n_i`$ of the experiment $`i`$. We select 3000 as an arbitrary lower bound and then divide the dataset of experiments into 2 groups: $`n_i >= 3000`$, and $`n_i < 3000`$.

Then, we have that:

- 97 experiments have 3000 jobs or more.
- 1941 experiments have less than 3000 jobs.

Here are their distributions:

$`n_i < 3000`$             |  $`n_i >= 3000`$
:-------------------------:|:-------------------------:
![experiments_u3000](https://earth.bsc.es/gitlab/es/autosubmit/uploads/d6354cb2dc01796bbd78f45afbed7653/experiments_u3000.jpg) |  ![experiments_o3000](https://earth.bsc.es/gitlab/es/autosubmit/uploads/f17f0c25240eba989d63aca1a32db604/experiments_o3000.jpg)

Seems that the vast majority of our experiments are under $10000$ jobs. Theoretically, Autosubmit GUI should be able to handle experiments around $20000$ jobs in less than 10 minutes; however, we do not really want to make our users wait ten minutes to see the graph or tree representation of his experiment. 

### Data comparison

Now let's take a look at how the distribution has been changing (or not) over time.

It was decided (as a simplification) to make this small analysis using the prefix of the experiment name, so we start by dividing our dataset into three groups according to the first 2 characters of the experiment name. We have that:

- 396 experiments starting with **a0**.
- 747 experiments starting with **a1**.
- 705 experiments starting with **a2**.

Now consider that:

- Experiments starting with **a0** date approximately from April 2015 until April 2018.
- Experiments starting with **a1** date approximately from April 2018 until July 2019. Around 16 months.
- Experiments starting with **a2** date approximately from July 2019 until April 2020. Around 9 months.

![experimentsa0](https://earth.bsc.es/gitlab/es/autosubmit/uploads/440dcf5d5004d30348ba6f0949d20e16/experimentsa0.jpg)

**a1**            |  **a2**
:-------------------------:|:-------------------------:
![experimentsa1](https://earth.bsc.es/gitlab/es/autosubmit/uploads/9a62156eeb3ce86203668caeccf98df2/experimentsa1.jpg) |  ![experimentsa2](https://earth.bsc.es/gitlab/es/autosubmit/uploads/2338ba5e63bd39a0a0a27fd70fe467b1/experimentsa2.jpg)

- There are 396 experiments starting with **a0**.
- There are 747 experiments starting with **a1**.
- There are 705 experiments starting with **a2**.

The distributions of **a1** and **a2** seem more closely related. So let's take a quick look at their distributions:

**a1** $`n_i < 3000`$      |  **a1** $`n_i >= 3000`$
:-------------------------:|:-------------------------:
![experimentsa1_u3000](https://earth.bsc.es/gitlab/es/autosubmit/uploads/2bb3a2408edf3a28f1ebfa77237ca3b0/experimentsa1_u3000.jpg) |  ![experimentsa1_o3000](https://earth.bsc.es/gitlab/es/autosubmit/uploads/5cb2063dc36a92d0bc4aa39e02d03155/experimentsa1_o3000.jpg)

- 44 experiments have 3000 jobs or more.
- 703 experiments have less than 3000 jobs.

**a2** $`n_i < 3000`$      |  **a2** $`n_i >= 3000`$
:-------------------------:|:-------------------------:
![experimentsa2_u3000](https://earth.bsc.es/gitlab/es/autosubmit/uploads/0227a575b1231b7bcda20862abaa851f/experimentsa2_u3000.jpg) |  ![experimentsa2_o3000](https://earth.bsc.es/gitlab/es/autosubmit/uploads/2269a284f8de0eb7f2b745858dc93585/experimentsa2_o3000.jpg)

- 47 experiments have 3000 jobs or more.
- 658 experiments have less than 3000 jobs.

### Conclusion

The prefix **a2** has been in usage for 7 months less than **a1** but it has almost the same number of experiments. Also, given the similarity of their distributions, it is not out of the realm of possibilities that the prefix **a2** will end with more experiments in $`n_i >= 3000`$ than **a1**. Of course, some prediction algorithm could give us an approximate number of large experiments that we should expect by the end of a2. However, this simple analysis has been performed only using the nomenclature difference of experiments. A more in-depth analysis should define appropriate time spans and filters to have a good starting point.

In conclusion, this small analysis serves as a way to justify the need for optimization in Autosubmit to handle large experiments (as measured by the number of jobs), and also to showcase some of the data that can be queried an analyzed now thanks to Autosubmit API.

Dataset (31-03-2020): [experiment_jobs.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/be0f0c16c8bc2ff4a30bfdf67b582488/experiment_jobs.txt)

## Development:

Let's start with an important question: 

**Should the database where the JobList is stored be individual for each experiment (like packages), or should it be centralized in a single database for all experiments (like ec_earth.db)?**

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 27, 2020, 18:53

It is a good question. By keeping them separate we will ensure the database do not grow too much, harming the performance. On the other hand, having n databases will make global requests longer.

In this case as we are talking about the workflow structure, I guess we will always look for a specific experiment, so having separate ddbb makes sense.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Mar 31, 2020, 18:37

Hi @mcastril @dbeltrankyl @pablohe 

I added an Analysis section that might be of interest. I have also attached the dataset I used.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 1, 2020, 16:37

Hi @wuruchi ,

This information is very valuable, thanks for the study! 

I think you should present this in the first CES meeting that we have teams presentations, and later in a seminar for all the department.

Some considerations and suggestions:

* I think that the general summary is so unbalanced to the small experiments because of the tests. I think you should exclude at least all the t0* experiment (or count them separately). There's also many tests with standard EXPID's, but I think it is ok just separating the t0*.

* The breakdown by prefix is very useful, and as you point out it supports our hypothesis that the demands are increasing. I guess you will have already ideas for metrics, but I will tell you some that I find they would be very useful:  
  * Experiment size averages for a0, a1, a2, in addition to the absolute values. You can also take separate mean values for the <3000 and >3000, to see how big is the typical small exp and typical big exp.
  * Experiments by day and jobs by day (raw measure, only counting total completed jobs per day maybe)? for a0, a1 and a2.
  * We also have to take into account that we have been deleting many new and old experiments, so you could maybe count the "created experiments". Just counting all the alphanumeric combinations from a000 to the current, for a0, a1 and a2. This is not exact because new experiments can overwrite deleted ones, but I think it only happens if you delete the last created experiment.
  *  I guess the peak of 2000 jobs is DustClim, or MONARCH-DA in general. I think you could do three categories, <1500, 1500-3000 and >3000.

Congratulations again for the quick report

@kserradell

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 2, 2020, 10:24

Thank you @mcastril. 

>I think you should present this in the first CES meeting that we have teams presentations, and later in a seminar for all the department.

That would be my pleasure.

>I think you should exclude at least all the t0\* experiment (or count them separately)

Yes, actually I forgot to exclude them in the initial filter.

>Experiment size averages for a0, a1, a2, in addition to the absolute values. You can also take separate mean values for the <3000 and >3000, to see how big is the typical small exp and typical big exp.

Sure

>Experiments by day and jobs by day (raw measure, only counting total completed jobs per day maybe)? for a0, a1 and a2.

Doing job level analysis might require some more effort, but it is doable.

>We also have to take into account that we have been deleting many new and old experiments, so you could maybe count the "created experiments". Just counting all the alphanumeric combinations from a000 to the current, for a0, a1 and a2. This is not exact because new experiments can overwrite deleted ones, but I think it only happens if you delete the last created experiment.

Currently, there is a way to know exactly the creation date of an experiment. We can use this to order the experiments properly.

>I guess the peak of 2000 jobs is DustClim, or MONARCH-DA in general. I think you could do three categories, <1500, 1500-3000 and >3000.

Sure. 

In general, I think we can include these changes before the CES presentation. I will set aside some time to start working on this analysis, as a secondary objective of this issue.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 2, 2020, 18:08

mentioned in commit 7068995161daed3dfd6a92d20f9a7ae86955db62

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 3, 2020, 12:11

Hello @mcastril @dbeltrankyl 

A first version of the JobList structure persistence implementation has been submitted for merge at !160.

The major features include:

* A `structure_[expid].db` file is created in the `pkl` folder of the experiment `[expid]`. Inside this file, a table `experiment_structure` is created.
* The experiment dependencies are stored in an edge list format: There is one row in the table `experiment_status` for each $`(u,v) \in E(G)`$ where $`E(G)`$ is the list of edges in the graph representation of the dependencies of the experiment.
* This file is created if it not exists.
* If the file exists but changes in the list of jobs have been detected, the table `experiment_status` is truncated and all dependencies inserted again. There are 2 main conditions to test if the job list has changed: the number of jobs in the stored structure should be equal to the number of jobs in the current generated list, and all the jobs in the current generated list should be present in the stored structure.
* If access to the structure file fails or the database engine fails, the process continues as usual.
* The file is generated for **all experiments independent of the number of jobs**.

It would be great to know your thoughts on this implementation.

Also, it has been already implemented in Autosubmit API, so users with large experiments should experience shorter loading times. The files created by Autosubmit API are under the user `webadmin`.

I forgot to mention: As a side consequence, the server `bscesweb04` should be experiencing far less computation workload. Is there a way to prove that?

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 3, 2020, 19:13

Hi @wuruchi and thanks for the details. I have a couple of questions:

* You say you are storing the edges in a database. Shouldn't you then use the number of edges as a measure for Autosubmit to detect that the experiment changed? Sometimes the structure is modified only to remove dependencies while maintaining the number of jobs. 

* Maybe there is a way to prune the database without truncating it. Maybe the difference if not worth for very large changes, but in case that only some dependencies are removed maybe an update is less costly. What do you think?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 6, 2020, 09:12

Hi @mcastril 

>You say you are storing the edges in a database. Shouldn't you then use the number of edges as a measure for Autosubmit to detect that the experiment changed? Sometimes the structure is modified only to remove dependencies while maintaining the number of jobs.

Yes, I will change the node count for edge count validation.

>Maybe there is a way to prune the database without truncating it. Maybe the difference if not worth for very large changes, but in case that only some dependencies are removed maybe an update is less costly. What do you think?

That could be too computationally costly. Considering that these databases only handle one experiment at the time, I think it is more efficient to insert all the edges again; although if it were a database that is constantly updated that wouldn't be a good approach.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 6, 2020, 10:19

> That could be too computationally costly. Considering that these databases only handle one experiment at the time, I think it is more efficient to insert all the edges again; although if it were a database that is constantly updated that wouldn't be a good approach. 

I am thinking in the limit cases. Let's say we have an experiment having 5000 jobs and the only change is removing the last job (supposing we have 2 key jobs at the end) that only depends on one job (this can perfectly happen, given that we have jobs that only perform a clean or a check at the end of the experiment). In that case it seems overkill to re-do the graph. 

Of course if re-doing the database takes only one second in every case (I assume that the main time is spent on the recalculation of the graph and not the DB operation itself) it is not worth to spend time on this, but if the time it's going to depend largely on the number of jobs maybe it's worth to take it into account, let's see...

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 8, 2020, 14:20

mentioned in commit 5b67995ea34d371340e4fea9257382401efc36b2

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jul 21, 2020, 11:52

closed
